### PR TITLE
[not ready] Enable per-segment speed updates just before contraction

### PR DESCRIPTION
**Pull request by danpat**
This change adds the ability to lookup per-segment (a segment is nodeA->nodeB) speed data just before graph contraction in `osrm-prepare`.

The main aim of this is to enable lookups of traffic data.  Generation of the data is left as an exercise to for the reader.

This change is not ready to merge yet, there are no tests.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -30,6 +30,7 @@ if(WIN32 AND MSVC_VERSION LESS 1800)
 endif()
 
 option(ENABLE_JSON_LOGGING "Adds additional JSON debug logging to the response" OFF)
+option(DEBUG_GEOMETRY "Enables an option to dump GeoJSON of the final routing graph" OFF)
```
Comment (TheMarex) [inline CMakeLists.txt]: Can we move the debug output to a different PR? Should be easy to separate by splitting the commit with `git rebase -i`.

File: algorithms/coordinate_calculation.hpp [hunk 2]
```diff
@@ -36,15 +36,15 @@ struct FixedPointCoordinate;
 namespace coordinate_calculation
 {
     double
-    great_circle_distance(const int lat1, const int lon1, const int lat2, const int lon2);
+    haversine_distance(const int lat1, const int lon1, const int lat2, const int lon2);
```
Comment (TheMarex) [inline algorithms/coordinate_calculation.hpp]: We we separate this out in an own commit as well?

File: contractor/contractor.hpp [hunk 3]
```diff
@@ -55,6 +55,14 @@ SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
 class Contractor
 {
+  public:
```
Comment (TheMarex) [inline contractor/contractor.hpp]: This is actually obsolete now. We can try to merge my PR on this first and then rebase.

File: contractor/processing_chain.cpp [hunk 4]
```diff
@@ -72,14 +75,22 @@ int Prepare::Run()
 
     DeallocatingVector<EdgeBasedEdge> edge_based_edge_list;
 
-    size_t max_edge_id = LoadEdgeExpandedGraph(config.edge_based_graph_path, edge_based_edge_list);
+    size_t max_edge_id = LoadEdgeExpandedGraph(config.edge_based_graph_path, edge_based_edge_list,
+            config.edge_segment_lookup_path,
+            config.edge_penalty_path,
+            config.segment_speed_lookup_path.string());
```
Comment (TheMarex) [inline contractor/processing_chain.cpp]: Why the `.string()` here? This should be consistent.

File: CMakeLists.txt (hunk 1 above)
Comment (danpat) [inline CMakeLists.txt]: Yup, it's hidden in the reflog, I'll cherry-pick it into `develop` right now.

File: contractor/processing_chain.cpp [hunk 5]
```diff
@@ -100,13 +115,46 @@ int Prepare::Run()
     return 0;
 }
 
+namespace std {
+
```
Comment (TheMarex) [inline contractor/processing_chain.cpp]: Can you reformat this section using the `.clangformat` file in the repository root?

File: algorithms/coordinate_calculation.hpp (hunk 2 above)
Comment (danpat) [inline algorithms/coordinate_calculation.hpp]: Yup, done.  It was in aa233eb which was buried in the reflog.

File: CMakeLists.txt (hunk 1 above)
Comment (danpat) [inline CMakeLists.txt]: I'll make a branch then remove this from this PR.  It depends on the edge_segment lookup loop in this PR, so it can't go into develop before.

File: contractor/contractor.hpp (hunk 3 above)
Comment (danpat) [inline contractor/contractor.hpp]: :+1:

File: contractor/processing_chain.cpp [hunk 6]
```diff
@@ -120,22 +168,154 @@ std::size_t Prepare::LoadEdgeExpandedGraph(
...
+        SimpleLogger().Write() << "Segment speed data supplied, will update edge weights from " << segment_speed_filename;
+        io::CSVReader<3> csv_in(segment_speed_filename);
+        csv_in.set_header("from_node","to_node","speed");
+        unsigned from_node_id; unsigned to_node_id; unsigned speed;
+        while (csv_in.read_row(from_node_id, to_node_id, speed)) {
+            segment_speed_lookup[std::pair<unsigned,unsigned>(from_node_id,to_node_id)] = speed;
+        }
+    }
+
+#ifdef DEBUG_GEOMETRY
```
Comment (TheMarex) [inline contractor/processing_chain.cpp]: Can we try to wrap this blocks in functions and move them to a file in `util/`?

File: contractor/processing_chain.cpp [hunk 7]
```diff
@@ -120,22 +168,154 @@ std::size_t Prepare::LoadEdgeExpandedGraph(
...
+
+                auto speed_iter = segment_speed_lookup.find(std::pair<unsigned, unsigned>(previous_osm_node_id, this_osm_node_id));
+                if (speed_iter != segment_speed_lookup.end())
+                {
+                    int new_segment_weight = std::max(1, static_cast<int>(std::floor(
+                                    (segment_length * 10.) / (speed_iter->second / 3.6)
+                                    + .5)));
+                    new_weight += new_segment_weight;
+
+#ifdef DEBUG_GEOMETRY
```
Comment (TheMarex) [inline contractor/processing_chain.cpp]: Same here. Needs to move out of the code.

File: contractor/contractor_options.cpp [hunk 8]
```diff
@@ -55,7 +55,10 @@ ContractorOptions::ParseArguments(int argc, char *argv[], ContractorConfig &cont
                          ->default_value(tbb::task_scheduler_init::default_num_threads()),
         "Number of threads to use")(
 		"core,k", boost::program_options::value<double>(&contractor_config.core_factor)
-						 ->default_value(1.0),"Percentage of the graph (in vertices) to contract [0.1]");
+						 ->default_value(1.0),"Percentage of the graph (in vertices) to contract [0..1]")(
```
Comment (daniel-j-h) [inline contractor/contractor_options.cpp]: [0,1] is never checked for, what happens if I say -42 or 9000? make this robust

File: contractor/contractor_options.hpp [hunk 9]
```diff
@@ -58,6 +61,9 @@ struct ContractorConfig
     //The remaining vertices form the core of the hierarchy 
     //(e.g. 0.8 contracts 80 percent of the hierarchy, leaving a core of 20%)
     double core_factor;
+
+    boost::filesystem::path segment_speed_lookup_path;
+
```
Comment (daniel-j-h) [inline contractor/contractor_options.hpp]: why is this a `b::fs::path` and the other paths above are string types?

File: contractor/processing_chain.cpp [hunk 10]
```diff
@@ -50,6 +52,8 @@ SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 #include <string>
 #include <thread>
 #include <vector>
+#include <exception>
+#include <ctime>
```
Comment (daniel-j-h) [inline contractor/processing_chain.cpp]: what do you need the ctime header for?

File: contractor/processing_chain.cpp [hunk 11]
```diff
@@ -100,12 +105,45 @@ int Prepare::Run()
...
+                                           const std::string &edge_segment_lookup_filename,
+                                           const std::string &edge_penalty_filename,
+                                           const std::string &segment_speed_filename)
 {
     SimpleLogger().Write() << "Opening " << edge_based_graph_filename;
-    boost::filesystem::ifstream input_stream(edge_based_graph_filename, std::ios::in | std::ios::binary);
+    boost::filesystem::ifstream input_stream(edge_based_graph_filename,
+                                             std::ios::in | std::ios::binary);
+
+    bool update_edge_weights = segment_speed_filename != "";
```
Comment (daniel-j-h) [inline contractor/processing_chain.cpp]: can be const

File: contractor/processing_chain.cpp [hunk 12]
```diff
@@ -100,12 +105,45 @@ int Prepare::Run()
...
+    boost::filesystem::ifstream edge_segment_input_stream;
+    boost::filesystem::ifstream edge_fixed_penalties_input_stream;
+
+    if (update_edge_weights)
+    {
+        edge_segment_input_stream.open(edge_segment_lookup_filename,
+                                       std::ios::in | std::ios::binary);
+        edge_fixed_penalties_input_stream.open(edge_penalty_filename,
+                                               std::ios::in | std::ios::binary);
+        if (!edge_segment_input_stream.is_open() || !edge_fixed_penalties_input_stream.is_open())
```
Comment (daniel-j-h) [inline contractor/processing_chain.cpp]: instead of checking for `is_open`, I would just use the implicit conversion to bool that also checks for failues, like:

    if (not stream) { /* oops */ }

File: contractor/processing_chain.cpp [hunk 13]
```diff
@@ -120,18 +158,85 @@ std::size_t Prepare::LoadEdgeExpandedGraph(
...
+                edge_segment_input_stream.read(reinterpret_cast<char *>(&this_osm_node_id),
+                                               sizeof(this_osm_node_id));
+                edge_segment_input_stream.read(reinterpret_cast<char *>(&segment_length),
+                                               sizeof(segment_length));
+                edge_segment_input_stream.read(reinterpret_cast<char *>(&segment_weight),
+                                               sizeof(segment_weight));
+
+                auto speed_iter = segment_speed_lookup.find(
+                    std::pair<unsigned, unsigned>(previous_osm_node_id, this_osm_node_id));
+                if (speed_iter != segment_speed_lookup.end())
```
Comment (daniel-j-h) [inline contractor/processing_chain.cpp]: you can simplify by testing for

    map.count(key) > 0

File: contractor/processing_chain.cpp [hunk 14]
```diff
@@ -120,18 +158,85 @@ std::size_t Prepare::LoadEdgeExpandedGraph(
...
+                {
+                    int new_segment_weight =
+                        std::max(1, static_cast<int>(std::floor(
+                                        (segment_length * 10.) / (speed_iter->second / 3.6) + .5)));
```
Comment (daniel-j-h) [inline contractor/processing_chain.cpp]: wait what is going on here? can not follow --- is this a heuristic?

File: contractor/processing_chain.cpp [hunk 15]
```diff
@@ -100,12 +105,45 @@ int Prepare::Run()
...
+
+    boost::filesystem::ifstream edge_segment_input_stream;
+    boost::filesystem::ifstream edge_fixed_penalties_input_stream;
+
+    if (update_edge_weights)
+    {
+        edge_segment_input_stream.open(edge_segment_lookup_filename,
+                                       std::ios::in | std::ios::binary);
+        edge_fixed_penalties_input_stream.open(edge_penalty_filename,
+                                               std::ios::in | std::ios::binary);
```
Comment (daniel-j-h) [inline contractor/processing_chain.cpp]: no need to specify `ios::in` for **i**fstream, throughout the codebase --- that's what the i in ifstream is for :)

File: extractor/edge_based_graph_factory.cpp [hunk 16]
```diff
@@ -437,6 +457,14 @@ void EdgeBasedGraphFactory::GenerateEdgeExpandedEdges(
...
                 {
                     distance += speed_profile.u_turn_penalty;
+                } 
+                if (generate_edge_lookup && turn_penalty > 0) 
+                {
+                    const QueryNode &v = m_node_info_list[node_v];
+
+                    const float bearing_uv = coordinate_calculation::bearing(first_coordinate,v);
+                    float uvw_normal = bearing_uv + turn_angle/2;
+                    while (uvw_normal >= 360.) { uvw_normal -= 360.; }
```
Comment (daniel-j-h) [inline extractor/edge_based_graph_factory.cpp]: did you just implement the modulo operator? :P

File: contractor/processing_chain.cpp (hunk 10 above)
Comment (danpat) [inline contractor/processing_chain.cpp]: Leftover from debugging, will remove.

File: contractor/contractor_options.hpp (hunk 9 above)
Comment (danpat) [inline contractor/contractor_options.hpp]: Good question :-)  Willfix.

File: contractor/processing_chain.cpp (hunk 13 above)
Comment (danpat) [inline contractor/processing_chain.cpp]: I actually use `speed_iter.second` a couple of lines lower.  If I used `.count()`, I'd still need to look it up.

**Comment by emiltin** [comment]
is there some description of how this is used and what format the input data should have?

**Comment by danpat** [comment]
@emiltin Not yet, but I'll put something on the wiki soon.

The TL;DR is:

  You can use this to update arbitrary road segment weights at the last possible moment in the processing pipeline before CH is applied.

  run `osrm-extract --generate-edge-lookup=yes` to create the lookup files necessary (`.osrm.edge_segment_lookup` and `.osrm.edge_penalties`).  These are exploded mappings between the edge-based edges and the node-based edges.

  then `osrm-prepare --segment-speed-file <something>.csv ....`

The `segment-speed-file` is a simple list of OSM segments and a speed value in km/h:

    237846,823742,32

and an OSM segment is any pair of nodes that are joined by a way to form a node-based edge.  Basically, you can pick any two directly connected nodes in OSM, and add them as an entry in the lookup file.

We're using this for traffic data: we map-match to discover the node ID pairs we're interested in, then update edge weights using those pairs just before we run `osrm-prepare`.

We do need to re-match whenever we import new OSM data, but so far, the churn rate has stayed fairly manageable.

**Comment by emiltin** [comment]
thanks. planning on adding cucumber tests?
